### PR TITLE
build: allow fallback for `CMAKE_Swift_COMPILER` migration

### PR DESCRIPTION
**Pull request by compnerd**
The original change c0d6d8f1907fd8b700b4dbece7f99301bd5ebe53 added a
soft migration path from `SWIFT_NATIVE_SWIFT_TOOLS_PATH` to
`CMAKE_Swift_COMPILER`, which broke some setups.  Subsequently
9da766582079223c749490fd07b36550358d1188 added an overly restrictive
check to ensure that `CMAKE_Swift_COMPILER` is always set when
`SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER` is set.  This loosens the check
to ensure that either `CMAKE_Swift_COMPILER` or
`SWIFT_NATIVE_SWIFT_TOOLS_PATH` can be used during the migration.

Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

Resolves SR-NNNN.

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @davezarzycki

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
Running a nightly build for Windows at https://dev.azure.com/compnerd/swift-build/_build/results?buildId=52716&view=results

**Review by davezarzycki — APPROVED** [review]
Verified. I restored my setup to how it was before #38123. That is to say, I changed my build scripts to no longer let Swift's CMake find the installed swift once again. I then applied #38238, this pull request, and did a clean build. Everything works. Thanks!

**Review by davezarzycki — CHANGES_REQUESTED** [review]
Wait, I forgot a restorative change. The bug still exists:
```
cd /home/dave/b/u/t/tools/swift/stdlib/public/core && /usr/bin/python3.9 /home/dave/ro_s/u/swift/utils/line-directive @/home/dave/b/u/t/tools/swift/stdlib/public/core/dfaab9867ab9868918d54c7626d620fe0cb9c79c.txt -- -c -sdk / -target x86_64-unknown-linux-gnu -resource-dir /home/dave/b/u/t/./lib/swift -O -D INTERNAL_CHECKS_ENABLED -D SWIFT_RUNTIME_OS_VERSIONING -module-cache-path /home/dave/b/u/t/./module-cache -no-link-objc-runtime -enable-library-evolution -Xfrontend -enforce-exclusivity=unchecked -nostdimport -parse-stdlib -module-name Swift -Xfrontend -group-info-path -Xfrontend /home/dave/ro_s/u/swift/stdlib/public/core/GroupInfo.json -swift-version 5 -runtime-compatibility-version none -disable-autolinking-runtime-compatibility-dynamic-replacements -warn-swift3-objc-inference-complete -Xfrontend -verify-syntax-tree -Xllvm -sil-inline-generics -Xllvm -sil-partial-specialization -Xfrontend -enable-experimental-concise-pound-file -Xfrontend -enable-ossa-modules -Xcc -DswiftCore_EXPORTS -warn-implicit-overrides -module-link-name swiftCore -whole-module-optimization -parse-as-library -I /home/dave/b/u/t/./lib/swift/linux -o /home/dave/b/u/t/tools/swift/stdlib/public/core/LINUX/x86_64/Swift.o @/home/dave/b/u/t/tools/swift/stdlib/public/core/dfaab9867ab9868918d54c7626d620fe0cb9c79c.txt
Traceback (most recent call last):
  File "/home/dave/ro_s/u/swift/utils/line-directive", line 742, in <module>
    run()
  File "/home/dave/ro_s/u/swift/utils/line-directive", line 694, in run
    command = subprocess.Popen(
  File "/usr/lib64/python3.9/subprocess.py", line 951, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib64/python3.9/subprocess.py", line 1821, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '-c'
```

**Comment by davezarzycki** [comment]
Weird, the following results in `NOTFOUND` and `FALSE` respectively but no error is generated:

```
if (SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER)
  if (NOT CMAKE_Swift_COMPILER AND NOT SWIFT_PREBUILT_SWIFT)
    message(ERROR "Swift Compiler not found and SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER is set")
  endif()
endif()
```

**Comment by compnerd** [comment]
Huh?  That sounds like the issue, but why?

**Comment by compnerd** [comment]
@davezarzycki - I added a horrible workaround to the CMake in the hopes that it evaluates properly.

File: CMakeLists.txt [hunk 1]
```diff
@@ -538,7 +538,7 @@ else()
 endif()
 
 if (SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER)
-  if (NOT CMAKE_Swift_COMPILER AND NOT SWIFT_PREBUILT_SWIFT)
+  if ("${CMAKE_Swift_COMPILER}" STREQUAL "CMAKE_Swift_COMPILER-NOTFOUND" AND NOT SWIFT_PREBUILT_SWIFT)
```
Comment (davezarzycki) [inline CMakeLists.txt]: In my testing and debugging, it needs to be: `"${CMAKE_Swift_COMPILER}" STREQUAL "NOTFOUND"`

**Review by davezarzycki — CHANGES_REQUESTED** [review]
So overall, after the `STREQUAL` fix, the net result is the same to me. I can't use `SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER` anymore unless I supply a pre-built Swift, which if that's the goal, then fine, I can live without `SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER`.

CC @gottesmm

File: CMakeLists.txt (hunk 1 above)
Comment (compnerd) [reply CMakeLists.txt]: Ah, okay.

**Comment by compnerd** [comment]
Wait, I'm not sure how the build is supposed to work in this case.  You are saying that CMake doesn't find the Swift compiler nor is `SWIFT_PREBUILT_SWIFT` set, so how is it supposed to find `swiftc`?  Is the expectation that it will find it in the path?

**Comment by davezarzycki** [comment]
One of the supported build modes uses the just built `swiftc` to compile swift code (see the top level CMakeLists.txt, line  578). But after @gottesmm's change, this mode became incompatible with `SWIFT_BUILD_RUNTIME_WITH_HOST_COMPILER`.

**Comment by compnerd** [comment]
Ah, for that I think that we should definitely get @gottesmm's opinion on introducing a more precise `SWIFT_BUILD_SWIFT_STANDARD_LIBRARY_WITH_HOST_COMPILER` instead of misattributing the compilation of the standard library to the runtime.  The runtime is technically below the standard library, and building that with the host compiler is different than building the stdlib with the host compiler.

**Comment by davezarzycki** [comment]
Right, I was allowing for a liberal definition of "runtime" or that the build variable name was no longer ideal but otherwise ya, we need @gottesmm's input.

**Comment by davezarzycki** [comment]
@gottesmm -- ping?

**Comment by davezarzycki** [comment]
@gottesmm -- ping?

**Comment by compnerd** [comment]
This is woefully out of date at this point, closing this off.  I imagine that is now longer valuable.

**Pull request closed without merging**